src/remote/remote_test.go

```diff
@@ -0,0 +1,271 @@
+	server.HighApiVersion.Major = 1
+	server.LowApiVersion.Major = 1
+	c := newClient()
+	assert.Error(t, c.CheckInitialised())
+	assert.Contains(t, c.CheckInitialised().Error(), "1.0.0 - 1.1.0")
+}
+
+func TestUnsupportedDigest(t *testing.T) {
+	defer server.Reset()
+	server.DigestFunction = []pb.DigestFunction_Value{
+		pb.DigestFunction_SHA256,
+		pb.DigestFunction_SHA384,
+		pb.DigestFunction_SHA512,
+	}
+	c := newClient()
+	assert.Error(t, c.CheckInitialised())
+}
+
+func TestStoreAndRetrieve(t *testing.T) {
+	c := newClient()
+	c.CheckInitialised()
+	target := core.NewBuildTarget(core.BuildLabel{PackageName: "package", Name: "target1"})
+	target.AddSource(core.FileLabel{File: "src1.txt", Package: "package"})
+	target.AddSource(core.FileLabel{File: "src2.txt", Package: "package"})
+	target.AddOutput("out1.txt")
+	// The hash is arbitrary as far as this package is concerned.
+	key, _ := hex.DecodeString("2dd283abb148ebabcd894b306e3d86d0390c82a7")
+	err := c.Store(target, key, []string{"plz-out/gen/package/out1.txt"})
+	assert.NoError(t, err)
+	// Remove the old file, but remember its contents so we can compare later.
+	contents, err := ioutil.ReadFile("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	err = os.Remove("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	// Now retrieve back the output of this thing.
+	err = c.Retrieve(target, key)
+	assert.NoError(t, err)
+	cachedContents, err := ioutil.ReadFile("plz-out/gen/package/out1.txt")
+	assert.NoError(t, err)
+	assert.Equal(t, contents, cachedContents)
+}
+
+func newClient() *Client {
+	config := core.DefaultConfiguration()
+	config.Build.Path = []string{"/usr/local/bin", "/usr/bin", "/bin"}
+	// Can't use NewDefaultBuildState since we need to modify the config first.
+	state := core.NewBuildState(1, nil, 4, config)
+	state.Config.Remote.URL = "127.0.0.1:9987"
+	return New(state)
+}
+
+// A testServer implements the server interface for the various servers we test against.
+type testServer struct {
+	DigestFunction                []pb.DigestFunction_Value
+	LowApiVersion, HighApiVersion semver.SemVer
+	actionResults                 map[string]*pb.ActionResult
+	blobs                         map[string][]byte
+	bytestreams                   map[string][]byte
+}
+
+func (s *testServer) GetCapabilities(ctx context.Context, req *pb.GetCapabilitiesRequest) (*pb.ServerCapabilities, error) {
+	return &pb.ServerCapabilities{
+		CacheCapabilities: &pb.CacheCapabilities{
+			DigestFunction: s.DigestFunction,
+			ActionCacheUpdateCapabilities: &pb.ActionCacheUpdateCapabilities{
+				UpdateEnabled: true,
+			},
+			MaxBatchTotalSizeBytes: 2048,
+		},
+		LowApiVersion:  &s.LowApiVersion,
+		HighApiVersion: &s.HighApiVersion,
+	}, nil
+}
+
+func (s *testServer) Reset() {
+	s.DigestFunction = []pb.DigestFunction_Value{
+		pb.DigestFunction_SHA1,
+		pb.DigestFunction_SHA256,
+	}
+	s.LowApiVersion = semver.SemVer{Major: 2}
+	s.HighApiVersion = semver.SemVer{Major: 2, Minor: 1}
```

Comment:
Ah, I see how this affected the first test now - feels like this might be neater in the construction of server, and done in a `setUp` method?

Reply:
`setUp` method? this isn't Python... :)